tokio/src/net/pipe.rs

```diff
@@ -486,16 +548,77 @@ impl Receiver {
     where
         P: AsRef<Path>,
     {
-        Receiver::open_internal(path.as_ref())
+        Receiver::open_internal(path.as_ref(), false)
     }
 
-    fn open_internal(path: &Path) -> io::Result<Receiver> {
+    /// Creates a new `Receiver` from a FIFO file in read-write access mode.
+    ///
+    /// This function will open the FIFO file at the specified path in `O_RDWR` access
+    /// mode and associate the pipe with the default event loop for reading.
+    ///
+    /// `Receiver` in read-write access mode can be used to implement resilient reading.
+    /// Unlike a `Receiver` opened in read-only mode with [`open`], pipe opened with
+    /// this function will not fail with `UnexpectedEof` during reading if the writing
+    /// end of the pipe closes the file.
+    ///
+    /// [`open`]: Self::open
+    ///
+    /// # Notes
+    ///
+    /// You should not use functions waiting for EOF such as [`read_to_end`] with
+    /// a `Receiver` in read-write access mode, since it may wait forever. `Receiver`
+    /// in this mode also holds an open writing end, which prevents receiving EOF.
+    ///
+    /// [`read_to_end`]: crate::io::AsyncReadExt::read_to_end
+    ///
+    /// # Errors
+    ///
+    /// Fails with any standard OS error if it occurs.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// use tokio::net::pipe;
+    /// use tokio::io::AsyncReadExt;
+    /// # use std::error::Error;
+    ///
+    /// const FIFO_NAME: &str = "path/to/a/fifo";
+    ///
+    /// # async fn dox() -> Result<(), Box<dyn Error>> {
+    /// let mut rx = pipe::Receiver::open_rw(FIFO_NAME)?;
+    /// loop {
+    ///     let mut msg = vec![0; 256];
+    ///     rx.read_exact(&mut msg).await?;
+    ///     /* handle message */
+    /// }
+    /// # }
+    /// ```
+    ///
+    /// # Panics
+    ///
+    /// This function panics if it is not called from within a runtime with
+    /// IO enabled.
+    ///
+    /// The runtime is usually set implicitly when this function is called
+    /// from a future driven by a tokio runtime, otherwise runtime can be set
+    /// explicitly with [`Runtime::enter`](crate::runtime::Runtime::enter) function.
+    #[cfg(target_os = "linux")]
+    #[cfg_attr(docsrs, doc(cfg(target_os = "linux")))]
+    pub fn open_rw<P>(path: P) -> io::Result<Receiver>
+    where
+        P: AsRef<Path>,
+    {
+        Receiver::open_internal(path.as_ref(), true)
+    }
+
+    fn open_internal(path: &Path, write_access: bool) -> io::Result<Receiver> {
         let file = std::fs::OpenOptions::new()
             .read(true)
+            .write(write_access)
             .custom_flags(libc::O_NONBLOCK)
             .open(path)?;
         let raw_fd = file.into_raw_fd();
-        // Safety: Raw fd was created a valid file.
+        // Safety: Raw fd was created from a valid file.
         let mio_rx = unsafe { mio_pipe::Receiver::from_raw_fd(raw_fd) };
         Receiver::from_mio(mio_rx)
```

Comment:
I think it's better to call `is_fifo` to verify that the `file` is indeed a pipe, then call `Receiver::from_file_unchecked` to avoid the `unsafe` code.